File: Components/DevExpressVCL/Demos/ExpressPivotGrid/CBuilder/ChartConnectionDemo/ChartConnectionDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ChartConnectionDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxDBPivotGrid"
#pragma link "cxClasses"
#pragma link "cxCustomData"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGrid"
#pragma link "cxGridChartView"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxPivotGridChartConnection"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxNavigator"
#pragma resource "*.dfm"
TfrmChartConnection *frmChartConnection;
//---------------------------------------------------------------------------
__fastcall TfrmChartConnection::TfrmChartConnection(TComponent* Owner)
        : TfrmDemoBasicMain(Owner)
{   
}

void __fastcall TfrmChartConnection::FormCreate(TObject* Sender)
{
  TfrmDemoBasicMain::FormCreate(Sender);  
  PivotGrid()->ApplyBestFit();
  InitializeDiagramList();
}

TcxCustomPivotGrid* __fastcall TfrmChartConnection::PivotGrid()
{
  return DBPivotGrid;
}

void __fastcall TfrmChartConnection::InitializeDiagramList()
{
  TcxGridChartDiagram *ADiagram;
  TMenuItem *AMenuItem;
  miChartType->Clear();
  for (int I = 0; I < cxGridChartView->DiagramCount; I++)
  {
	ADiagram = cxGridChartView->Diagrams[I];
	AMenuItem = new TMenuItem(miChartType);
	AMenuItem->AutoCheck = True;
	AMenuItem->RadioItem = True;
	AMenuItem->Checked = ADiagram == cxGridChartView->ActiveDiagram;
	AMenuItem->OnClick = miChartColumDiagramClick;
	AMenuItem->Caption = ADiagram->DisplayText;
	AMenuItem->Tag = I;
	miChartType->Add(AMenuItem);
  }
}

//---------------------------------------------------------------------------
void __fastcall TfrmChartConnection::pgfPaymentTypeGetGroupImageIndex(
	  TcxPivotGridField *Sender, const TcxPivotGridViewDataItem *AItem,
	  int &AImageIndex, TAlignment &AImageAlignHorz,
	  TcxAlignmentVert &AImageAlignVert)
{
  AnsiString Card = ((TcxPivotGridViewDataItem *)AItem)->Value;
  if (SameText(Card, "Cash")) AImageIndex = 0;
  else if (SameText(Card, "AmEx")) AImageIndex = 1;
	   else if (SameText(Card, "Master")) AImageIndex = 2;
			else if (SameText(Card, "Visa")) AImageIndex = 3;
}
//---------------------------------------------------------------------------
void __fastcall TfrmChartConnection::miVisibleCellsClick(TObject *Sender)
{
  cxPivotGridChartConnection->SourceData = (TcxPivotGridChartViewSourceData)((TMenuItem*)Sender)->Tag;
}
void __fastcall TfrmChartConnection::miSourceColumnClick(TObject *Sender)
{
  cxPivotGridChartConnection->SourceForCategories = TcxPivotGridChartViewSourceForCategories(((TMenuItem*)Sender)->Tag);

}
void __fastcall TfrmChartConnection::miChartColumDiagramClick(TObject *Sender)
{
  cxGridChartView->ActiveDiagram = cxGridChartView->Diagrams[((TMenuItem*)Sender)->Tag];
}

void __fastcall TfrmChartConnection::cxGridChartViewActiveDiagramChanged(TcxGridChartView *Sender,
	TcxGridChartDiagram *ADiagram)
{
	InitializeDiagramList();
}

File: Components/DevExpressVCL/Demos/ExpressPivotGrid/Delphi/ChartConnectionDemo/ChartConnectionDemoMain.pas
unit ChartConnectionDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, Variants, Forms, Dialogs,
  SysUtils, Menus, StdCtrls, Controls, cxLookAndFeels, DemoBasicMain,
  cxControls, cxCustomPivotGrid, cxDBPivotGrid, DemoBasicDM, cxStyles,
  cxCustomData, cxGraphics, cxFilter, cxData, cxDataStorage, cxEdit, DB, cxDBData,
  Classes, cxClasses, cxPivotGridChartConnection, cxGridLevel,
  cxGridCustomView, cxGridCustomTableView, cxGridTableView,
  cxGridDBTableView, cxGrid, cxGridChartView;

type
  TfrmChartConnection = class(TfrmDemoBasicMain)
    DBPivotGrid: TcxDBPivotGrid;
    pgfPurchaseDate: TcxDBPivotGridField;
    pgfPaymentType: TcxDBPivotGridField;
    pgfQuantity: TcxDBPivotGridField;
    pgfCarName: TcxDBPivotGridField;
    pgfUnitPrice: TcxDBPivotGridField;
    pgfCompanyName: TcxDBPivotGridField;
    pgfPaymentAmount: TcxDBPivotGridField;
    cxPivotGridChartConnection: TcxPivotGridChartConnection;
    cxGridDBTableView1: TcxGridDBTableView;
    cxGrid: TcxGrid;
    cxGridLevel: TcxGridLevel;
    cxGridChartView: TcxGridChartView;
    miChartOptions: TMenuItem;
    miSourceData: TMenuItem;
    miVisibleCells: TMenuItem;
    miSelectedCells: TMenuItem;
    miSourceForCategories: TMenuItem;
    miSourceRow: TMenuItem;
    miSourceColumn: TMenuItem;
    N3: TMenuItem;
    miChartType: TMenuItem;
    procedure FormCreate(Sender: TObject);
    procedure pgfPaymentTypeGetGroupImageIndex(Sender: TcxPivotGridField;
      const AItem: TcxPivotGridViewDataItem; var AImageIndex: Integer;
      var AImageAlignHorz: TAlignment;
      var AImageAlignVert: TcxAlignmentVert);
    procedure miVisibleCellsClick(Sender: TObject);
    procedure miSourceColumnClick(Sender: TObject);
    procedure miChartColumDiagramClick(Sender: TObject);
    procedure cxGridChartViewActiveDiagramChanged(Sender: TcxGridChartView;
      ADiagram: TcxGridChartDiagram);
  private
  protected
    function GetPivotGrid: TcxCustomPivotGrid; override;
    procedure InitializeDiagramList;
  end;

var
  frmChartConnection: TfrmChartConnection;

implementation

{$R *.dfm}

uses
  DemoUtils;

function TfrmChartConnection.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := DBPivotGrid;
end;

procedure TfrmChartConnection.InitializeDiagramList;
var
  ADiagram: TcxGridChartDiagram;
  AMenuItem: TMenuItem;
  I: Integer;
begin
  miChartType.Clear;
  for I := 0 to cxGridChartView.DiagramCount - 1 do
  begin
    ADiagram := cxGridChartView.Diagrams[I];
    AMenuItem := TMenuItem.Create(miChartType);
    AMenuItem.AutoCheck := True;
    AMenuItem.RadioItem := True;
    AMenuItem.Checked := ADiagram = cxGridChartView.ActiveDiagram;
    AMenuItem.OnClick := miChartColumDiagramClick;
    AMenuItem.Caption := ADiagram.DisplayText;
    AMenuItem.Tag := Integer(ADiagram);
    miChartType.Add(AMenuItem);
  end;
end;

procedure TfrmChartConnection.FormCreate(Sender: TObject);
begin
  inherited;
  PivotGrid.ApplyBestFit;
  InitializeDiagramList;
end;

procedure TfrmChartConnection.pgfPaymentTypeGetGroupImageIndex(
  Sender: TcxPivotGridField; const AItem: TcxPivotGridViewDataItem;
  var AImageIndex: Integer; var AImageAlignHorz: TAlignment;
  var AImageAlignVert: TcxAlignmentVert);
var
  Card: string;
begin
  Card := VarToStr(AItem.Value);
  if SameText(Card, 'AmEx') then
    AImageIndex := 1
  else
    if SameText(Card, 'Cash') then
      AImageIndex := 0
    else
      if SameText(Card, 'Master') then
        AImageIndex := 2
      else
        if SameText(Card, 'Visa') then
          AImageIndex := 3;
end;

procedure TfrmChartConnection.miVisibleCellsClick(Sender: TObject);
begin
  cxPivotGridChartConnection.SourceData :=
    TcxPivotGridChartViewSourceData(TComponent(Sender).Tag);
end;

procedure TfrmChartConnection.miSourceColumnClick(Sender: TObject);
begin
  cxPivotGridChartConnection.SourceForCategories :=
    TcxPivotGridChartViewSourceForCategories(TComponent(Sender).Tag);
end;

procedure TfrmChartConnection.miChartColumDiagramClick(Sender: TObject);
begin
  cxGridChartView.ActiveDiagram := TcxGridChartDiagram(TComponent(Sender).Tag);
end;

procedure TfrmChartConnection.cxGridChartViewActiveDiagramChanged(
  Sender: TcxGridChartView; ADiagram: TcxGridChartDiagram);
var
  AItem: TObject;
begin
  AItem := dxDemoFindMenuItem(mmMain, 'miChartType');
  dxDemoCheckedItemByTag(AItem, Integer(cxGridChartView.ActiveDiagram));  
end;

end.
